sourcemap/fleet_store.go

```diff
@@ -71,35 +79,74 @@ func newFleetStore(
 	fleetCfg *config.Fleet,
 	cfgs []config.SourceMapMetadata,
 ) (fleetStore, error) {
-	// TODO(stn): Add support for multiple fleet hosts
-	// cf. https://github.com/elastic/apm-server/issues/5514
-	host := fleetCfg.Hosts[0]
-	fleetURLs := make(map[key]string)
+	sourceMapURLs := make(map[key]string)
+	fleetBaseURLs := make([]string, len(fleetCfg.Hosts))
 
 	for _, cfg := range cfgs {
 		k := key{cfg.ServiceName, cfg.ServiceVersion, cfg.BundleFilepath}
-		u, err := common.MakeURL(fleetCfg.Protocol, cfg.SourceMapURL, host, 8220)
+		sourceMapURLs[k] = cfg.SourceMapURL
+	}
+
+	for i, host := range fleetCfg.Hosts {
+		baseURL, err := common.MakeURL(fleetCfg.Protocol, "", host, defaultFleetPort)
 		if err != nil {
 			return fleetStore{}, err
 		}
-		fleetURLs[k] = u
+		fleetBaseURLs[i] = baseURL
 	}
+
 	return fleetStore{
-		apikey:    "ApiKey " + fleetCfg.AccessAPIKey,
-		fleetURLs: fleetURLs,
-		c:         c,
+		apikey:        "ApiKey " + fleetCfg.AccessAPIKey,
+		fleetBaseURLs: fleetBaseURLs,
+		sourceMapURLs: sourceMapURLs,
+		c:             c,
+		rng:           rand.New(rand.NewSource(time.Now().UnixNano())),
 	}, nil
 }
 
 func (f fleetStore) fetch(ctx context.Context, name, version, path string) (string, error) {
 	k := key{name, version, path}
-	fleetURL, ok := f.fleetURLs[k]
+	sourceMapURL, ok := f.sourceMapURLs[k]
 	if !ok {
 		return "", fmt.Errorf("unable to find sourcemap.url for service.name=%s service.version=%s bundle.path=%s",
 			name, version, path,
 		)
 	}
 
+	idxs := f.rng.Perm(len(f.fleetBaseURLs))
+	respReceived := make(chan string, 1)
+	fetchErrReturned := make(chan error, 1)
+
+	for idx := range idxs {
+		go func(i int) {
+			baseURL := f.fleetBaseURLs[i]
+			fleetURL := baseURL + sourceMapURL
+			resp, err := sendRequest(f, ctx, fleetURL)
+			if err != nil {
+				fetchErrReturned <- err
+			} else {
+				respReceived <- resp
+			}
+		}(idx)
+	}
+
+	for range idxs {
+		select {
+		case resp := <-respReceived:
+			return resp, nil
+		case err := <-fetchErrReturned:
+			// If the error is not a query failure
+			if !strings.Contains(err.Error(), errMsgFleetFailure) {
+				return "", err
+			}
+		}
```

Comment:
The way this is written, we'll return the first response, but a later response might be successful.

Reply:
Also, if the first response received is successful, we could cancel any remaining requests that are still in-flight.